language/move-vm/runtime/src/interpreter.rs

```diff
@@ -180,7 +177,8 @@ impl<'txn> Interpreter<'txn> {
                     )?;
                     let func = resolver.function_at(fh_idx);
                     if func.is_native() {
-                        self.call_native(&resolver, context, func, vec![])?;
+                        let caller = current_frame.function.module_id();
```

Comment:
I have a case of using a native function where you need to know the calling module.  I have a module similar in functionality to the event module. And the function must know the module from which it is called. Without an explicit indication of the type in code. 
I have a module A, B, C, Store, and a script. The Store module contains a method that will save some resources using calling the native store function. There is a white list of modules from the context of which it is possible to do this, and prohibits the call of Store::store_resources from the script.

Reply:
I have to say I find passing the knowledge of the calling module too much information for a native. But I suppose that depends on what you need to do.
I am curious about your plans to use native functions with the MoveVM. 
I mean which part of the VM are planning to get visibility on and to build what.
I mean, at the end that is knowledge that you can expose explicitly in certain scenarios and if you cannot I am not sure what the developer expectation should be (how do I know what happens and why)

Reply:
I share the concerns about needing to know the calling module. The permissions model of Move is intentionally designed to avoid the need to do a stack walk for security checks.

> I have a module A, B, C, Store, and a script. The Store module contains a method that will save some resources using calling the native store function. There is a white list of modules from the context of which it is possible to do this, and prohibits the call of Store::store_resources from the script.

There are design patterns that can be used to do this without exposing the calling module. The one we usually use is something like
```
native fun store_resources(some_data: u64, capability: &StoreResourcesCapability); 

public fun grant_capability(): StoreResourcesCapability {
  if (// sender is in some whitelist of accepted addresses) {
    move_to_sender(StoreResourcesCapability{})
  }
}


// some other module that you want to allow to call store_resources
module Caller {
  resource T { cap: StoreResourcesCapability }

  public publish() {
    move_to_sender<t> { cap: StoreResources::grant_capability() };
  }

  public call(some_data: u64) {
    StoreResources::store_resources(some_data, &borrow_global<T>.cap)
  }
}

```

This way, it won't be possible for anyone to call `store_resources` unless they can get access to a `&StoreResourcesCapability`. Since `grant_capability` restricts access to the addresses you specify, you can ensure it can't be called  from anywhere you don't expect.